File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Java/perf_counters/perfVolLatency.java
/**
 * $Id:$
 *
 * perfVolLatency.java
 *
 * Copyright (c) 2009 NetApp, Inc. All rights reserved.
 * Specifications subject to change without notice.
 *
 *
 * This Sample code is supported from DataFabric Manager 3.7.1
 * onwards.
 * However few of the functionalities of the sample code may
 * work on older versions of DataFabric Manager.
 */

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

import netapp.manage.NaAPIFailedException;
import netapp.manage.NaElement;
import netapp.manage.NaServer;

public class perfVolLatency {
    public static void print_help() {
        System.out.println("\nCommand:");
        System.out.println("java perfVolLatency <dfm> <user> <password> "
                + "<aggr_name>");
        System.out.println("<dfm>       -- DFM Server name");
        System.out.println("<user>      -- User name");
        System.out.println("<password>  -- Password");
        System.out.println("<aggr-name> -- Name of the aggregate in format "
                + "storage:aggrName");
        System.out.println("-----------------------------------------------"
                + "---------------------\n");
        System.out.println("This sample code prints average latency of all "
                + "the volumes ");
        System.out.println("present in the given aggregate. This data can be"
                + "used to ");
        System.out.println("generate distribution chart for volume average"
                + "latency. ");
        System.out.println("To generate the graph, redirect output of this"
                + "sample code to");
        System.out.println("an Excel sheet.");

        System.exit(1);
    }

    /**
     * getVolLatency : Calculates the volume latency for all the volumes in the
     * aggregate specified.
     * 
     * @param server
     *            DFM server that is contacted to retrieve the information
     * @param aggr
     *            Parent aggregate, latency of volumes under this aggregate are
     *            calculated
     * @return This method outputs the values by itself, no return values
     */

    public static void getVolLatency(NaServer server, String aggr) {
        NaServer connectedServer = server;
        String targetAggr = aggr;
        NaElement perfApiInput, instanceInfo, counterInfo, perfObjectCounter;
        NaElement perfApiOutput;
        String volumeName;
        String localTime;

        /*
         * Construct the API to be called here. (perf-get-counter-data)
         * Specified : the duration for api to execute, number of samples and
         * the method of collecting data.
         */

        perfApiInput = new NaElement("perf-get-counter-data");
        perfApiInput.addNewChild("duration", "6000");
        perfApiInput.addNewChild("number-samples", "50");
        perfApiInput.addNewChild("time-consolidation-method", "average");

        instanceInfo = new NaElement("instance-counter-info");
        instanceInfo.addNewChild("object-name-or-id", targetAggr);

        counterInfo = new NaElement("counter-info");

        perfObjectCounter = new NaElement("perf-object-counter");
        perfObjectCounter.addNewChild("object-type", "volume");
        perfObjectCounter.addNewChild("counter-name", "avg_latency");

        counterInfo.addChildElem(perfObjectCounter);
        instanceInfo.addChildElem(counterInfo);
        perfApiInput.addChildElem(instanceInfo);

        try {
            perfApiOutput = connectedServer.invokeElem(perfApiInput);

            NaElement iterValue;
            boolean generateTimeArray = true;
            ArrayList timeArray = new ArrayList();
            ArrayList dataArray = new ArrayList();
            int nSamples = 0;
            int i = 0;
            String[] counterArray;

            List instances = perfApiOutput.getChildByName("perf-instances")
                    .getChildren();

            Iterator instanceIter = instances.iterator();

            /*
             * Populate the counter data for volume latency.
             */
            System.out.print("Time");
            while (instanceIter.hasNext()) {

                iterValue = (NaElement) instanceIter.next();
                System.out.print("\t");

                if (iterValue.getChildByName("instance-name") != null) {
                    volumeName = iterValue.getChildContent("instance-name");
                    System.out.print(volumeName);
                }

                if (iterValue.getChildByName("counters") != null) {
                    List counters1 = iterValue.getChildByName("counters")
                            .getChildren();
                    Iterator counterIter1 = counters1.iterator();

                    while (counterIter1.hasNext()) {

                        NaElement counter1 = (NaElement) counterIter1.next();
                        String counterData = counter1
                                .getChildContent("counter-data");
                        if (counterData == null || counterData.length() == 0)
                            break;
                        nSamples += counterData.split(",").length;
                        counterArray = counterData.split(",");

                        for (i = 0; i < counterArray.length; i++) {
                            String[] counterDataArray = new String[counterArray[i]
                                    .split(":").length];

                            counterDataArray = counterArray[i].split(":");

                            if (generateTimeArray == true) {
                                localTime = new SimpleDateFormat(
                                        "dd/MM/yyyy HH:mm:ss")
                                        .format(new Date(
                                                Long.parseLong(counterDataArray[0]) * 1000));

                                timeArray.add(localTime);

                            }
                            dataArray.add(new Double(Double
                                    .parseDouble(counterDataArray[1])));
                        }

                        generateTimeArray = false;

                    }
                }

            }

            int a, k;

            String[] timeSamples = new String[timeArray.size()];
            timeArray.toArray(timeSamples);
            Double[] dataSamples = new Double[dataArray.size()];
            dataArray.toArray(dataSamples);

            /*
             * Print data for volume latencies.
             */

            System.out.println("");
            for (a = 0; a < timeSamples.length; a++) {
                System.out.print(timeSamples[a] + "\t");
                for (k = a; k < dataSamples.length; k = k + timeSamples.length) {
                    System.out.print(dataSamples[k].toString() + "\t");
                }
                System.out.println("");
            }

        } catch (NaAPIFailedException exception) {
            exception.printStackTrace();
            System.exit(1);
        } catch (Exception exception) {
            exception.printStackTrace();
            System.exit(1);
        }
    }

    /**
     * @param args
     */
    public static void main(String[] args) {

        NaServer server;

        if (args.length < 4) {
            print_help();
        }

        String aggr = new String(args[3]);

        try {
            server = new NaServer(args[0], 1, 0);
            server.setStyle(NaServer.STYLE_LOGIN_PASSWORD);
            server.setAdminUser(args[1], args[2]);
            server.setServerType(NaServer.SERVER_TYPE_DFM);
            server.setPort(8088);

            getVolLatency(server, aggr);

        } catch (Exception exception) {
            exception.printStackTrace();
            System.exit(1);
        }
    }

}

File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Ruby/perf_counters/perf_vol_latency.rb
#===============================================================#
#                                                               #
# ID                                                            #
#                                                               #
# perf_vol_latency.rb                                           #
#                                                               #
# Copyright (c) 2011 NetApp, Inc. All rights reserved.          #
# Specifications subject to change without notice.              #
#                                                               #
# This Sample code is supported from DataFabric Manager 3.7.1   #
# onwards.                                                      #
# However few of the functionalities of the sample code may     #
# work on older versions of DataFabric Manager.                 #
#===============================================================#

$:.unshift '../../../../../../../lib/ruby/NetApp'
require 'NaServer'


def print_help() 
	print ("\nCommand:\n")
	print ("perf_vol_latency.rb <dfm> <user> <password> <aggr-name>\n")
	print ("<dfm>       -- DFM Server name\n")
	print ("<user>      -- User name\n")
	print ("<password>  -- Password\n")
	print ("<aggr-name> -- Name of the aggregate in format storage:aggrName\n")
	print ("-" * 80 + "\n")
	print ("This sample code prints average latency of all the volumes \n")
	print ("present in the given aggregate. This data can be used to \n")
	print ("generate distribution chart for volume average latency. \n")
	print ("To generate the graph, redirect output of this sample code to\n")
	print ("an Excel sheet.\n")
	exit 
end


def invoke_perf_zapi() 
	dfm = ARGV[0]
	user = ARGV[1]
	pw  = ARGV[2]
	# Initialize server context
	server_ctx = NaServer.new(dfm, 1, 0)
	server_ctx.set_admin_user(user, pw)
	server_ctx.set_server_type("DFM")
	# Create API request
	perf_in = NaElement.new("perf-get-counter-data")
	perf_in.child_add_string("duration", 6000)
	perf_in.child_add_string("number-samples", 50)
	perf_in.child_add_string("time-consolidation-method", "average")
	instance_info = NaElement.new("instance-counter-info")
	instance_info.child_add_string("object-name-or-id", ARGV[3])
	counter_info = NaElement.new("counter-info")
	perf_obj_ctr = NaElement.new("perf-object-counter")
	perf_obj_ctr.child_add_string("object-type", "volume")
	perf_obj_ctr.child_add_string("counter-name", "avg_latency")
	counter_info.child_add(perf_obj_ctr)
	instance_info.child_add(counter_info)
	perf_in.child_add(instance_info)
	# Invoke API
	perf_out = server_ctx.invoke_elem(perf_in)	
	if(perf_out.results_status() == "failed") 
		print(perf_out.results_reason() +"\n")
		exit
	end	
	return perf_out
end


def extract_perf_counter_data(perf_out)
	gen_time_arr = 1
	i = 0
	instance = perf_out.child_get("perf-instances")
	instances = instance.children_get()
	instances.each do |rec|
		vol_name = rec.child_get_string("instance-name")		
		if (vol_name) 
			$vol_arr.push(vol_name)
		end		
		counters = rec.child_get("counters")
		perf_cnt_data = counters.children_get()		
		perf_cnt_data.each do |rec1|
			counter_str = rec1.child_get_string("counter-data")
			# counter-data is in time1:val1,time2:val2,..,timen:valn format.
			# Extract data from this format.
			counter_arr = counter_str.split(',')			
			counter_arr.each do |time_val|
				time_val_arr = time_val.split(':')
				if(gen_time_arr) 
					$time_arr[i] = Time.at(time_val_arr[0].to_f)
				end		
				$data_arr[i] = time_val_arr[1]
				i = i + 1
			end			
			gen_time_arr = nil
		end
	end
end


def print_output() 
	j = 0
	k = 0
	iter = 0
	nsamples = $time_arr.length 
	data_arr_len = $data_arr.length
	print ("Time\t\t\t\t")
	$vol_arr.each do |vol_name|
		print(vol_name + "\t")
	end	
	print ("\n")
	while(iter < nsamples)
		print($time_arr[iter].asctime + "\t")			
		k = iter
		while(k < data_arr_len)
			print ($data_arr[k].to_s + "\t\t")
			k = k + nsamples
		end		
		print "\n"
		iter = iter + 1
	end
end


def main() 		
	perf_out = invoke_perf_zapi()
	extract_perf_counter_data(perf_out)
	print_output()
	exit 
end


# Command line arguments
args = ARGV.length
# check for valid number of parameters
if (args != 4)	
	print_help() 
end
$time_arr = []
$data_arr = []
$vol_arr = []
#Invoke routine
main()

'''#=========================== POD ============================#

=head1 NAME

  perf_vol_latency.rb - Get average latency of all volumes present in an
	aggregate.


=head1 SYNOPSIS

  perf_vol_latency.rb <dfm> <user> <password> <aggr-name>

=head1 ARGUMENTS

  <dfm>
  DFM Server name.

  <user>
  User name.

  <password>
  Password.

  <aggr-name>
  Name of the aggregate in format storage:aggrName.

=head1 SEE ALSO

  NaElement.rb, NaServer.rb

=head1 COPYRIGHT

 Copyright (c) 2009 NetApp, Inc. All rights reserved.
 Specifications subject to change without notice.

=cut'''

